### PR TITLE
[AutoDiff upstream] Add stdlib `Differentiable` conformances and derivatives.

### PR DESCRIPTION
**Pull request by dan-zheng**
Add `Differentiable` conformances and derivatives for stdlib types and functions to the `_Differentiation` module. Most upstreaming work done by @ematejska!

---

### `Differentiable` conformances
- `FloatingPoint`-conforming types: `Float16`, `Float`, `Double`, `Float80`
- `Array` where `Element: Differentiable`
- `SIMD`-conforming types where `Scalar: Differentiable & BinaryFloatingPoint, Scalar.TangentVector: BinaryFloatingPoint`

### Derivatives
- Floating-point operations: `+`, `+=`, `-`, `-=`, `*`, `*=`, `/`, `/=`
- Array operations: `Array.init(repeating:count:)`, `Array.subscript.get`, `Array.+(_:_:)`
- `SIMD` operations: `+`, `-`, `*`, `/`
- Most tgmath operations: `exp`, `log`, `cos`, `sin`, etc.

Differentiable collection higher-order functions:
- `Array.differentiableMap(_:)`, `Array.differentiableReduce(_:_:)`

Add end-to-end derivative tests.

---

Todos:

- `Differentiable.zeroTangentVector` protocol requirement and implementations.
  - TF-1238: to be upstreamed.
- `SIMD.sum` differentiation.
  - TF-1103: `@_alwaysEmitIntoClient` derivative registration.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Comment by dan-zheng** [comment]
Two tests fail with `-O` due to SR-12548: `SILCombiner::visitPartialApplyInst` SIL verification failure. I'll submit a fix soon.

File: stdlib/public/Differentiation/FloatingPointDifferentiation.swift.gyb [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Swift
+import SwiftShims
```
Comment (ematejska) [inline stdlib/public/Differentiation/FloatingPointDifferentiation.swift.gyb:14]: We might not need to import SwiftShims on master. What do you think?  Aside from that looks good.

**Review by ematejska — APPROVED** [review]

**Review by ematejska — COMMENTED** [review]
Looks like there are a couple of test failures still.

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
Test failures due to SR-12548 fixed in https://github.com/apple/swift/pull/30877.
@swift-ci Please test

**Comment by dan-zheng** [comment]
Fixed test failures for 32-bit platforms.
@swift-ci Please test

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Comment by dan-zheng** [comment]
@swift-ci Please test macOS

**Comment by dan-zheng** [comment]
@swift-ci Please test Windows

**Pull request merged**